livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,384 @@
+        self,
+        opts: STTOptions,
+        api_key: str,
+        http_session: aiohttp.ClientSession,
+        num_channels: int = 1,
+        max_retry: int = 32,
+    ) -> None:
+        super().__init__()
+
+        self._opts = opts
+        self._num_channels = num_channels
+        self._api_key = api_key
+        self._speaking = False
+        self._session = http_session
+        self._queue = asyncio.Queue[Union[rtc.AudioFrame, str]]()
+        self._event_queue = asyncio.Queue[Optional[stt.SpeechEvent]]()
+        self._closed = False
+        self._max_retry = max_retry
+
+        if self._num_channels != 1:
+            raise ValueError(f"AssemblyAI only supports mono audio, but a `num_channels` of {self._num_channels} was provided")
+
+        # keep a list of final transcripts to combine them inside the END_OF_SPEECH event
+        self._final_events: List[stt.SpeechEvent] = []
+
+    def push_frame(self, frame: rtc.AudioFrame) -> None:
+        if self._closed:
+            raise ValueError("cannot push frame to closed stream")
+
+        self._queue.put_nowait(frame)
+
+    async def aclose(self, *, wait: bool = True) -> None:
+        self._closed = True
+        self._queue.put_nowait(SpeechStream._CLOSE_MSG)
+
+        if not wait:
+            self._main_task.cancel()
+
+        with suppress(asyncio.CancelledError):
+            await self._main_task
+
+        await self._session.close()
+
+    async def _main_task(self) -> None:
+        return self._run(self._max_retry)
+
+    async def _run(self, max_retry: int) -> None:
+        """
+        Run a single websocket connection to AssemblyAI and make sure to reconnect
+        when something went wrong.
+        """
+        try:
+            retry_count = 0
+            while not self._closed:
+                try:
+                    live_config = {
+                        "sample_rate": self._opts.sample_rate,
+                        "word_boost": self._opts.word_boost,
+                        "encoding": self._opts.encoding,
+                        "disable_partial_transcripts": self._opts.disable_partial_transcripts,
+                        "enable_extra_session_information": self._opts.enable_extra_session_information,
+                    }
+
+                    headers = {
+                        "Authorization": self._api_key,
+                        "Content-Type": "application/json",
+                    }
+
+                    ws_url = "wss://api.assemblyai.com/v2/realtime/ws"
+                    url = f"{ws_url}?{urlencode(live_config).lower()}"
+                    ws = await self._session.ws_connect(url, headers=headers)
+                    retry_count = 0  # connected successfully, reset the retry_count
+
+                    await self._run_ws(ws)
+                except Exception:
+                    # Something went wrong, retry the connection
+                    if retry_count >= max_retry:
+                        print(
```

Comment:
Could we use `logging.Logger`s instead so we can get structured logging, job_id, etc for better debugging? (here and below, there are a few prints floating around)

Reply:
fixed ✅ thanks!